### PR TITLE
Support unstable moves via stable in unstable items

### PR DESCRIPTION
**Pull request by yaahc**
part of https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/moving.20items.20to.20core.20unstably and a blocker of https://github.com/rust-lang/rust/pull/90328. 

The libs-api team needs the ability to move an already stable item to a new location unstably, in this case for Error in core. Otherwise these changes are insta-stable making them much harder to merge.

This PR attempts to solve the problem by checking the stability of path segments as well as the last item in the path itself, which is currently the only thing checked.

### CODE REVIEW

**Comment by yaahc** [comment]
somewhat stuck on this, could use some guidance, context posted here: https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/how.20does.20the.20deprecation.20logic.20check.20parent.20modules.3F

File: src/test/ui/lint/lint-stability.rs
Comment (yaahc) [inline src/test/ui/lint/lint-stability.rs:198]: I was surprised these test cases were already present. I don't think we have any examples of either of these kinds of situations currently that we'll be breaking by changing this behavior, but I'm curious if there was any explicit justification for this behavior back when these test cases were added.

File: library/std/src/lib.rs
Comment (yaahc) [inline library/std/src/lib.rs:295]: necessary because of `impl std::error::Error for alloc::ffi::NulError` in std.

File: library/std/src/panic.rs
Comment (yaahc) [inline library/std/src/panic.rs:14]: necessary because panic items are invoked via `$crate::rt` and `rt` has it's own unstable feature that wasn't previously explicitly allowed

File: library/core/src/unicode/mod.rs
Comment (yaahc) [inline library/core/src/unicode/mod.rs:1]: Necessary because `UNICODE_VERSION` and therefore `core::unicode::UNICODE_VERSION` is already stable.

**Comment by yaahc** [comment]
This is now also a T-libs-api PR because of the `core::unicode` stabilization :/

**Comment by yaahc** [comment]
Also blocked on https://github.com/rust-lang/stdarch/pull/1303

@rustbot blocked

File: library/std/src/lib.rs
Comment (yaahc) [inline library/std/src/lib.rs:215]: necessary because of https://github.com/rust-lang/rust/blob/f5cdb567f12e707331445386177c8c24b88af856/library/std/src/panicking.rs#L408-L409

**Comment by yaahc** [comment]
This should be ready now that https://github.com/rust-lang/stdarch/pull/1303 has merged.

@rustbot ready

**Comment by yaahc** [comment]
removing libs-api label and reverting changes to `core::unicode` stability based on this conversation: https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/accidental.20.60core.3A.3Aunicode.60.20stabilization

We decided we're going to intentionally break this and see what happens. There is a stable re-export via `core::char::UNICODE_VERSION` which seemed to be the original intentional export location, and no visible usages of `UNICODE_VERSION` though `core::unicode` on github codesearch, so we're hoping that nobody has depended on this unintentional stabilization and it is safe to break. If crater comes back with significant breakage or anyone comes back and complains about the breakage we can reconsider and will likely stabilize `core::unicode`.

**Comment by CAD97** [comment]
Just a drive-by note: `std::prelude::rust_2024` will have the same issue of stable-in-unstable without this patch

File: src/test/ui/lint/lint-stability.rs
Comment (cjgillot) [reply src/test/ui/lint/lint-stability.rs:198]: Is this still true?  I don't see this stabilization in the PR.

File: compiler/rustc_passes/src/stability.rs
Comment (cjgillot) [inline compiler/rustc_passes/src/stability.rs:794]: IIUC, we only report unstability in the path segments if the path itself is stable.  Could you add this to the comment?

**Review by cjgillot — APPROVED** [review]
This implementation looks good and quite elegant.
Left a few nits.
r=me if the impact on libs API is clear.

File: src/test/ui/lint/lint-stability.rs
Comment (yaahc) [reply src/test/ui/lint/lint-stability.rs:198]: Ah, no, I removed the stabilization, we decided to break it because we don't think anyone is depending on it, pending crater run

File: src/test/ui/stability-attribute/allow-through-unstable-misuse.rs
Comment (yaahc) [inline src/test/ui/stability-attribute/allow-through-unstable-misuse.rs:8]: I based my attribute loosely on `#[rustc_promotable]` and noticed in the process that it is exempted from the requirement to add ui tests for every error code. Rather than just cargo cult that I decided to try testing mine but I'm running into an issue where it emits the diagnostic twice...

Comment (yaahc) [reply src/test/ui/stability-attribute/allow-through-unstable-misuse.rs]: actually reversed that decision for now so I can focus on the rest of the PR, will revisit this once I'm happy with everything else.

**Comment by yaahc** [comment]
Didn't mean to initiate a perf test so removed those labels

**Comment by yaahc** [comment]
no, dang it rust timer, stop, please :sob:

File: src/test/ui/stability-attribute/rustc-encodable-stability.rs
Comment (yaahc) [inline src/test/ui/stability-attribute/rustc-encodable-stability.rs:3]: This doesn't seem to be finding rustc_serialize, so I'm not really sure how to interact with this deprecated API in order to test it from a ui test.

Comment (eddyb) [reply src/test/ui/stability-attribute/rustc-encodable-stability.rs:3]: I think moving the test from `ui` to `ui-fulldeps` is the simplest approach that can work (since it's a compiler crate). Otherwise you'd need to mock it or something.

Comment (eddyb) [inline src/test/ui/stability-attribute/rustc-encodable-stability.rs:3]: Nit: there should probably be a one-line gap here (and honestly I wish we had this as a lint), I sleepily misread this as the attribute being applied to the `extern crate` instead of the whole crate.

**Comment by Dylan-DPC** [comment]
this pr is no longer blocked since stdarch pr is merged

this is waiting on some changes by the author

**Comment by yaahc** [comment]
For some reason I keep running into a failure of `src/test/codegen/mem-replace-direct-memcpy.rs` locally which I doubt I broke so I'm pushing for now to see what CI says

**Comment by yaahc** [comment]
Ended up removing the ui-fulldeps test because I could not figure out how to make the test with `RustcEncodable/Decodable` compile and I given that they're already deprecated and have zero test coverage I expect it doesn't really matter and I'm unlikely to have broken anything that's not already broken...

**Comment by eddyb** [comment]
cc @petrochenkov I have a vague memory of there being some previous attempt at "check every segment of a path, not just the final one", but I'm not sure it was for stability or something else, or what the overall outcome was.

**Comment by yaahc** [comment]
> @yaahc so this PR led to a performance regression - I see that a perf run was performed (accidentally) before this was merged, but it wasn't looked at. Is there a reason that the performance difference wasn't taken into account?

only because I hadn't intended to run it in the first place, I just didn't pay attention to it.

> In either case, we should at least address the performance regression with a justification if one is warranted.
> 
> The change seems to be localized to incremental compilation scenarios. It seems queries like `incr_comp_encode_dep_graph` are being hit more often. Is this simply because of the additional pass that's happening or is there some other explanation I don't see?

I don't have a solid explanation for why that's happening, I don't have much experience with interpreting perf reports for the compiler. Where are you seeing that `incr_comp_encode_dep_graph` is being hit more often?

From the impl itself I'd expect that the main difference is that we hit the stability check function much more often because of: https://github.com/rust-lang/rust/pull/95956/files#diff-5f57ad10e1bdde3d046b258d390fd2ecc6f1511158aa130cebb72093da16ef29R851, but I don't know how that relates to incremental compilation at all.

**Comment by rylev** [comment]
You can see differences in query perf [on this page](https://perf.rust-lang.org/detailed-query.html?commit=24699bcbadff59693a89a5184afed87f6416cdea&base_commit=f1a8854f9be2e5cad764d630a53d26c7b72f8162&benchmark=diesel-1.4.8-check&scenario=incr-patched:%20println). You can see there that `incr_comp_encode_dep_graph` was executed more often. 

I'm also unsure, as I don't see anything that would cause encoding the dep graph more often, but I'm really not familiar enough with this code to say for sure. 

The regressions are sort of right on the line between being small enough to shrug off and big enough to really care about investigating. I'll leave it to you on what to do next. Feel free to reach out if you want to chat through things.

**Comment by nnethercote** [comment]
I ran a Cachegrind diff. The results absolutely match up with what @rylev said about the incremental dep graph.
```
--------------------------------------------------------------------------------
Ir
--------------------------------------------------------------------------------
72,523,795 (100.0%)  PROGRAM TOTALS

--------------------------------------------------------------------------------
Ir                  file:function
--------------------------------------------------------------------------------
1,988,457 ( 2.74%)  compiler/rustc_query_system/src/dep_graph/graph.rs:<rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::try_mark_previous_green::<rustc_query_impl::plumbing::QueryCtxt>
1,759,692 ( 2.43%)  library/core/src/num/uint_macros.rs:<rustc_data_structures::stable_hasher::StableHasher>::finish::<rustc_data_structures::fingerprint::Fingerprint>
1,537,026 ( 2.12%)  compiler/rustc_serialize/src/leb128.rs:<rustc_query_system::dep_graph::serialized::SerializedDepGraph<rustc_middle::dep_graph::dep_node::DepKind> as rustc_serialize::serialize::Decodable<rustc_serialize::opaque::MemDecoder>>::decode
1,455,664 ( 2.01%)  compiler/rustc_serialize/src/leb128.rs:<(rustc_query_system::dep_graph::serialized::SerializedDepNodeIndex, rustc_query_impl::on_disk_cache::AbsoluteBytePos) as rustc_serialize::serialize::Decodable<rustc_serialize::opaque::MemDecoder>>::decode
1,368,740 ( 1.89%)  compiler/rustc_middle/src/dep_graph/mod.rs:<rustc_middle::dep_graph::dep_node::DepKind as rustc_query_system::dep_graph::DepKind>::read_deps::<<rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::read_index::{closure#0}>
1,260,818 ( 1.74%)  compiler/rustc_query_system/src/query/plumbing.rs:rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<rustc_span::def_id::DefId, core::option::Option<rustc_attr::builtin::Stability>>>
1,225,791 ( 1.69%)  github.com-1ecc6299db9ec823/hashbrown-0.12.0/src/raw/mod.rs:<rustc_middle::dep_graph::dep_node::DepKind as rustc_query_system::dep_graph::DepKind>::read_deps::<<rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::read_index::{closure#0}>
1,155,286 ( 1.59%)  compiler/rustc_data_structures/src/sip128.rs:<rustc_data_structures::stable_hasher::StableHasher>::finish::<rustc_data_structures::fingerprint::Fingerprint>
1,091,722 ( 1.51%)  github.com-1ecc6299db9ec823/hashbrown-0.12.0/src/raw/mod.rs:rustc_query_system::query::plumbing::try_get_cached::<rustc_middle::ty::context::TyCtxt, rustc_query_system::query::caches::DefaultCache<rustc_span::def_id::DefId, core::option::Option<rustc_attr::builtin::Stability>>, core::option::Option<rustc_attr::builtin::Stability>, rustc_middle::ty::query::copy<core::option::Option<rustc_attr::builtin::Stability>>>
1,015,419 ( 1.40%)  compiler/rustc_query_system/src/query/plumbing.rs:rustc_query_system::query::plumbing::try_get_cached::<rustc_middle::ty::context::TyCtxt, rustc_query_system::query::caches::DefaultCache<rustc_span::def_id::DefId, core::option::Option<rustc_attr::builtin::Stability>>, core::option::Option<rustc_attr::builtin::Stability>, rustc_middle::ty::query::copy<core::option::Option<rustc_attr::builtin::Stability>>>
  995,121 ( 1.37%)  compiler/rustc_serialize/src/leb128.rs:<[rustc_query_system::dep_graph::graph::DepNodeIndex] as rustc_serialize::serialize::Encodable<rustc_serialize::opaque::FileEncoder>>::encode
  969,062 ( 1.34%)  github.com-1ecc6299db9ec823/smallvec-1.8.1/src/lib.rs:<smallvec::SmallVec<[rustc_query_system::dep_graph::graph::DepNodeIndex; 8]> as core::iter::traits::collect::Extend<rustc_query_system::dep_graph::graph::DepNodeIndex>>::extend::<core::iter::adapters::map::Map<core::slice::iter::Iter<rustc_query_system::dep_graph::serialized::SerializedDepNodeIndex>, <rustc_query_system::dep_graph::graph::CurrentDepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::promote_node_ad_deps_to_current::{closure#0}>>
  958,608 ( 1.32%)  compiler/rustc_serialize/src/leb128.rs:<[(rustc_query_system::dep_graph::serialized::SerializedDepNodeIndex, rustc_query_impl::on_disk_cache::AbsoluteBytePos)] as rustc_serialize::serialize::Encodable<rustc_query_impl::on_disk_cache::CacheEncoder>>::encode
  944,089 ( 1.30%)  compiler/rustc_query_system/src/dep_graph/serialized.rs:<rustc_query_system::dep_graph::serialized::GraphEncoder<rustc_middle::dep_graph::dep_node::DepKind>>::send
  923,416 ( 1.27%)  compiler/rustc_query_system/src/query/plumbing.rs:rustc_query_system::query::plumbing::get_query::<rustc_query_impl::queries::lookup_stability, rustc_query_impl::plumbing::QueryCtxt>
  923,104 ( 1.27%)  compiler/rustc_query_impl/src/on_disk_cache.rs:<rustc_query_impl::on_disk_cache::OnDiskCache>::try_load_query_result::<core::option::Option<rustc_attr::builtin::Stability>>
  826,887 ( 1.14%)  compiler/rustc_middle/src/ty/query.rs:<rustc_passes::stability::Checker as rustc_hir::intravisit::Visitor>::visit_path
  816,784 ( 1.13%)  compiler/rustc_query_system/src/query/plumbing.rs:rustc_query_system::query::plumbing::try_load_from_disk_and_cache_in_memory::<rustc_query_impl::plumbing::QueryCtxt, rustc_span::def_id::DefId, core::option::Option<rustc_attr::builtin::Stability>>
  745,584 ( 1.03%)  github.com-1ecc6299db9ec823/hashbrown-0.12.0/src/raw/mod.rs:<hashbrown::raw::RawTable<(rustc_query_system::dep_graph::serialized::SerializedDepNodeIndex, rustc_query_impl::on_disk_cache::AbsoluteBytePos)>>::insert::<hashbrown::map::make_hasher<rustc_query_system::dep_graph::serialized::SerializedDepNodeIndex, rustc_query_system::dep_graph::serialized::SerializedDepNodeIndex, rustc_query_impl::on_disk_cache::AbsoluteBytePos, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>
```

**Pull request merged**
